# Pandas For Data Analytics and EDA 

#### 25 important funtions:


In [234]:
import pandas as pd

In [235]:
df = pd.read_csv(r"C:\Users\Faizan_Shakeel\Desktop\ML Engineer\Learning1\Pandas_Newyork_city\Drug_driving.csv")

In [236]:
df[4000:4004]

stop_date stop_time  county_name driver_gender  driver_age_raw  \
4000  2006-02-15     15:10          NaN             M          1979.0   
4001  2006-02-15     15:25          NaN             M          1968.0   
4002  2006-02-15     15:25          NaN             M          1962.0   
4003  2006-02-15     15:30          NaN             M          1954.0   

      driver_age driver_race                   violation_raw  violation  \
4000        27.0       White  Equipment/Inspection Violation  Equipment   
4001        38.0       White                        Speeding   Speeding   
4002        44.0       White                        Speeding   Speeding   
4003        52.0       White                        Speeding   Speeding   

      search_conducted search_type stop_outcome is_arrested stop_duration  \
4000             False         NaN     Citation       False      0-15 Min   
4001             False         NaN     Citation       False      0-15 Min   
4002             False         NaN     Citation       False     16-30 Min   
4003             False         NaN     Citation       False      0-15 Min   

      drugs_related_stop  
4000               False  
4001               False  
4002               False  
4003               False

In [237]:
df.shape

(91741, 15)

# 1.  df.info()

In [238]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           91741 non-null  object 
 1   stop_time           91741 non-null  object 
 2   county_name         0 non-null      float64
 3   driver_gender       86406 non-null  object 
 4   driver_age_raw      86414 non-null  float64
 5   driver_age          86120 non-null  float64
 6   driver_race         86408 non-null  object 
 7   violation_raw       86408 non-null  object 
 8   violation           86408 non-null  object 
 9   search_conducted    91741 non-null  bool   
 10  search_type         3196 non-null   object 
 11  stop_outcome        86408 non-null  object 
 12  is_arrested         86408 non-null  object 
 13  stop_duration       86408 non-null  object 
 14  drugs_related_stop  91741 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 9.3+

In [239]:
df.isnull().sum()

stop_date                 0
stop_time                 0
county_name           91741
driver_gender          5335
driver_age_raw         5327
driver_age             5621
driver_race            5333
violation_raw          5333
violation              5333
search_conducted          0
search_type           88545
stop_outcome           5333
is_arrested            5333
stop_duration          5333
drugs_related_stop        0
dtype: int64

In [240]:
df.drop(["county_name"], axis=1,inplace=True)

# 5.  to_datetime()
So, while reading a CSV file, the DateTime objects in the file are read as string objects and therefore, it’s a little difficult to perform DateTime operations like time difference on a string. So, this is where the pandas “to_datetime()” method comes into play. You can provide various formats as per your requirement.# 

In [241]:
import datetime as dt

df["stop_date"] = pd.to_datetime(df["stop_date"])
df["stop_time"] = pd.to_datetime(df["stop_time"],format="%H:%M").dt.time

In [242]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   stop_date           91741 non-null  datetime64[ns]
 1   stop_time           91741 non-null  object        
 2   driver_gender       86406 non-null  object        
 3   driver_age_raw      86414 non-null  float64       
 4   driver_age          86120 non-null  float64       
 5   driver_race         86408 non-null  object        
 6   violation_raw       86408 non-null  object        
 7   violation           86408 non-null  object        
 8   search_conducted    91741 non-null  bool          
 9   search_type         3196 non-null   object        
 10  stop_outcome        86408 non-null  object        
 11  is_arrested         86408 non-null  object        
 12  stop_duration       86408 non-null  object        
 13  drugs_related_stop  91741 non-null  bool      

In [243]:
df.describe().T

count         mean         std   min     25%     50%  \
driver_age_raw  86414.0  1970.491228  110.914909   0.0  1967.0  1980.0   
driver_age      86120.0    34.011333   12.738564  15.0    23.0    31.0   

                   75%     max  
driver_age_raw  1987.0  8801.0  
driver_age        43.0    99.0

In [244]:
df["driver_race"].value_counts()

White       62158
Black       12244
Hispanic     9507
Asian        2259
Other         240
Name: driver_race, dtype: int64

In [245]:
df["search_type"].value_counts()

Incident to Arrest                                          1219
Probable Cause                                               891
Inventory                                                    220
Reasonable Suspicion                                         197
Protective Frisk                                             161
Incident to Arrest,Inventory                                 129
Incident to Arrest,Probable Cause                            106
Probable Cause,Reasonable Suspicion                           75
Incident to Arrest,Inventory,Probable Cause                   34
Incident to Arrest,Protective Frisk                           33
Probable Cause,Protective Frisk                               33
Inventory,Probable Cause                                      22
Incident to Arrest,Reasonable Suspicion                       13
Inventory,Protective Frisk                                    11
Incident to Arrest,Inventory,Protective Frisk                 11
Protective Frisk,Reasonab

In [246]:
df["stop_duration"].value_counts()

0-15 Min     69543
16-30 Min    13635
30+ Min       3228
2                1
1                1
Name: stop_duration, dtype: int64

# 10. fillna()

By “isnull()” and “sum()” checks the missing values in dataset.
we can see that this “driver_gender” feature has 5335 nan or missing values. 

So, we will fill up the missing values using the:
##### mode(a value that appears most frequently in a data set) of this particular feature using the “fillna()” function.

df["driver_gender"].isnull().sum()

In [247]:
df["driver_gender"].value_counts()

M    62895
F    23511
Name: driver_gender, dtype: int64

In [248]:
df["driver_gender"].fillna(df["driver_gender"].mode,inplace=True)

In [249]:
df["driver_gender"].isnull().sum()

0

###### Note – This is NOT THE most effective way of filling missing/NAN values. 
###### There are other much more efficient ways for imputation. There is a lot of thought which goes behind when we are working with missing values. But, this is just for explaining the functionality of the fillna() function.

# 12.    nunique()
- Generally, it is used in the case of categorical features to identify the no. of categories in a particular feature.
- We can use the “nunique()” function to find the no. of unique values in our series or data frame.

In [250]:
df['violation_raw'].nunique()

12

In [251]:
df['violation_raw'].value_counts()

Speeding                            48463
Other Traffic Violation             16224
Equipment/Inspection Violation      11020
Registration Violation               3432
Seatbelt Violation                   2952
Special Detail/Directed Patrol       2455
Call for Service                     1298
Violation of City/Town Ordinance      211
Motorist Assist/Courtesy              203
APB                                    79
Suspicious Person                      56
Warrant                                15
Name: violation_raw, dtype: int64

#     nsmallest() & nlargest()
- As the name suggests, “nsmallest() & nlargest()” functions are used to obtain “n” no. of rows from our dataset which are lowest or highest respectively.

In [252]:
df.nsmallest(5,'driver_age')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
10500 2006-09-30  01:30:00             M          1991.0        15.0   
17771 2007-06-11  12:30:00             M          1992.0        15.0   
18357 2007-07-04  00:40:00             F          1992.0        15.0   
25294 2008-04-21  11:19:00             M          1993.0        15.0   
45988 2010-11-10  19:39:00             F          1995.0        15.0   

      driver_race            violation_raw         violation  \
10500       Black  Other Traffic Violation  Moving violation   
17771       White                 Speeding          Speeding   
18357       White  Other Traffic Violation  Moving violation   
25294    Hispanic  Other Traffic Violation  Moving violation   
45988       White  Other Traffic Violation  Moving violation   

       search_conducted         search_type   stop_outcome is_arrested  \
10500             False                 NaN  Arrest Driver        True   
17771             False                 NaN       Citation       False   
18357             False                 NaN  Arrest Driver        True   
25294              True  Incident to Arrest  Arrest Driver        True   
45988             False                 NaN       Citation       False   

      stop_duration  drugs_related_stop  
10500       30+ Min               False  
17771      0-15 Min               False  
18357      0-15 Min               False  
25294       30+ Min               False  
45988     16-30 Min               False

In [253]:
df.nlargest(5,'driver_age')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
71539 2013-09-08  09:50:00             F          1914.0        99.0   
75037 2014-02-14  10:21:00             M          1920.0        94.0   
76956 2014-04-25  16:40:00             M          1924.0        90.0   
74348 2014-01-18  13:07:00             F          1925.0        89.0   
9829  2006-09-06  12:15:00             M          1918.0        88.0   

      driver_race                   violation_raw         violation  \
71539       White         Other Traffic Violation  Moving violation   
75037       White         Other Traffic Violation  Moving violation   
76956       White                        Speeding          Speeding   
74348       White  Equipment/Inspection Violation         Equipment   
9829        White                        Speeding          Speeding   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
71539             False         NaN     Citation       False     16-30 Min   
75037             False         NaN     Citation       False      0-15 Min   
76956             False         NaN      Warning       False      0-15 Min   
74348             False         NaN          N/D       False      0-15 Min   
9829              False         NaN     Citation       False     16-30 Min   

       drugs_related_stop  
71539               False  
75037               False  
76956               False  
74348               False  
9829                False

# groupby()
- The “groupby()” function is very useful in data analysis as 
- it allows us to unveil the underlying relationships among different variables. 
- we can apply Aggregations as well on the groups with the “agg()” function and pass it with various aggregation operations such as mean, size, sum, std etc.



In [254]:
import numpy as np

df.groupby('driver_race').agg(np.mean)

C:\Users\Faizan_Shakeel\AppData\Local\Temp\ipykernel_16824\693768411.py:3: FutureWarning: The operation <function mean at 0x000001FD7D00B250> failed on a column. If any error is raised, this will raise an exception in a future version of pandas. Drop these columns to avoid this warning.
  df.groupby('driver_race').agg(np.mean)


driver_age_raw  driver_age  search_conducted  drugs_related_stop
driver_race                                                                  
Asian           1971.271802   33.579672          0.022576            0.004869
Black           1970.870549   33.110519          0.064521            0.016416
Hispanic        1974.886072   31.565369          0.061428            0.013253
Other           1967.025000   34.548117          0.012500            0.004167
White           1969.792239   34.576958          0.028444            0.007658

# get_group()

- “get_group()” function to select a specific group.

In [255]:
df.groupby('driver_race').get_group('Asian')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
10    2005-07-13  16:20:00             M          1979.0        26.0   
20    2005-07-28  12:37:00             F          1979.0        26.0   
37    2005-09-30  14:00:00             M          1962.0        43.0   
67    2005-10-01  23:25:00             M          1974.0        31.0   
96    2005-10-03  02:20:00             F          1950.0        55.0   
...          ...       ...           ...             ...         ...   
91502 2015-12-18  20:18:00             M          1988.0        27.0   
91510 2015-12-18  23:37:00             M          1993.0        22.0   
91539 2015-12-20  09:27:00             M          1962.0        53.0   
91612 2015-12-26  09:07:00             M          1942.0        73.0   
91704 2015-12-30  10:56:00             M          1989.0        26.0   

      driver_race                   violation_raw  violation  \
10          Asian                        Speeding   Speeding   
20          Asian                        Speeding   Speeding   
37          Asian                        Speeding   Speeding   
67          Asian                        Speeding   Speeding   
96          Asian                        Speeding   Speeding   
...           ...                             ...        ...   
91502       Asian                        Speeding   Speeding   
91510       Asian                        Speeding   Speeding   
91539       Asian                        Speeding   Speeding   
91612       Asian                        Speeding   Speeding   
91704       Asian  Equipment/Inspection Violation  Equipment   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
10                False         NaN     Citation       False      0-15 Min   
20                False         NaN     Citation       False      0-15 Min   
37                False         NaN     Citation       False      0-15 Min   
67                False         NaN     Citation       False     16-30 Min   
96                False         NaN     Citation       False      0-15 Min   
...                 ...         ...          ...         ...           ...   
91502             False         NaN     Citation       False      0-15 Min   
91510             False         NaN     Citation       False      0-15 Min   
91539             False         NaN     Citation       False      0-15 Min   
91612             False         NaN     Citation       False      0-15 Min   
91704             False         NaN     Citation       False      0-15 Min   

       drugs_related_stop  
10                  False  
20                  False  
37                  False  
67                  False  
96                  False  
...                   ...  
91502               False  
91510               False  
91539               False  
91612               False  
91704               False  

[2259 rows x 14 columns]

### Note – We can combine various methods of pandas as per our requirement for a better understanding of data as shown below:

In [256]:
df.groupby("driver_race").driver_age.describe()

count       mean        std   min   25%   50%   75%   max
driver_race                                                             
Asian         2253.0  33.579672  11.688393  17.0  24.0  31.0  41.0  82.0
Black        12197.0  33.110519  11.493603  15.0  24.0  30.0  40.0  85.0
Hispanic      9477.0  31.565369  10.561938  15.0  23.0  29.0  38.0  76.0
Other          239.0  34.548117  11.607951  18.0  25.0  32.0  41.5  74.0
White        61949.0  34.576958  13.252149  15.0  23.0  31.0  44.0  99.0

# df.loc[] and df.iloc[]

- loc() and iloc() methods are used in slicing data from the pandas DataFrame which helps in filtering the data according to some given condition.

- loc – select by labels

- iloc – select by positions 

In [257]:
df.head(3)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   
1 2005-01-18  08:15:00             M          1965.0        40.0       White   
2 2005-01-23  23:15:00             M          1972.0        33.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   
2      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False

In [258]:
df.loc[34:41,["driver_race", "violation_raw"]]

driver_race                   violation_raw
34       White                        Speeding
35       White  Equipment/Inspection Violation
36       White                        Speeding
37       Asian                        Speeding
38       White                        Speeding
39       White                        Speeding
40       White  Equipment/Inspection Violation
41       White  Equipment/Inspection Violation

In [259]:
df.loc[(df.driver_age<16) & (df.violation=='Speeding')]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
17771 2007-06-11  12:30:00             M          1992.0        15.0   

      driver_race violation_raw violation  search_conducted search_type  \
17771       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
17771     Citation       False      0-15 Min               False

In [260]:
# iloc() slices the data frame in the specified rows and column range.

df.iloc[:3,4:7]

driver_age driver_race violation_raw
0        20.0       White      Speeding
1        40.0       White      Speeding
2        33.0       White      Speeding

# Sorting

- We can sort our DataFrame by index or values with Pandas
- sort_index()  
- sort_values() functions. 


In [261]:
df[['driver_gender','driver_age','driver_race','violation','is_arrested']].sort_values(by='driver_age')[9990:9995]

driver_gender  driver_age driver_race         violation is_arrested
36325             M        21.0       White         Equipment       False
74543             M        21.0       White  Moving violation       False
74069             M        21.0       White          Speeding       False
31725             M        21.0       Black          Speeding       False
20016             M        21.0       White          Speeding       False

In [262]:
df.sort_index(ascending = False, kind='quicksort', axis=1, na_position='last')[3:5]

#tried using Shift+tab

violation_raw violation stop_time   stop_outcome stop_duration  \
3  Call for Service     Other  17:15:00  Arrest Driver     16-30 Min   
4          Speeding  Speeding  10:00:00       Citation      0-15 Min   

   stop_date search_type  search_conducted is_arrested  drugs_related_stop  \
3 2005-02-20         NaN             False        True               False   
4 2005-03-14         NaN             False       False               False   

  driver_race driver_gender  driver_age_raw  driver_age  
3       White             M          1986.0        19.0  
4       White             F          1984.0        21.0

# Query
- query() function to filter our data frame as per our conditions or requirements as shown below:

In [263]:
df.query('80<driver_age>85')[:4]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
9829  2006-09-06  12:15:00             M          1918.0        88.0   
23590 2008-02-04  14:40:00             M          1922.0        86.0   
23847 2008-02-15  14:34:00             M          1922.0        86.0   
24898 2008-04-05  10:07:00             M          1922.0        86.0   

      driver_race violation_raw violation  search_conducted search_type  \
9829        White      Speeding  Speeding             False         NaN   
23590       White      Speeding  Speeding             False         NaN   
23847       White      Speeding  Speeding             False         NaN   
24898       White      Speeding  Speeding             False         NaN   

      stop_outcome is_arrested stop_duration  drugs_related_stop  
9829      Citation       False     16-30 Min               False  
23590     Citation       False      0-15 Min               False  
23847     Citation       False      0-15 Min               False  
24898     Citation       False      0-15 Min               False

In [264]:
df.query('75<driver_age<80 and driver_race=="Black"')

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
42160 2010-05-06  03:48:00             M          1932.0        78.0   
74882 2014-02-08  09:39:00             M          1936.0        78.0   
87208 2015-07-07  03:05:00             F          1937.0        78.0   
91154 2015-12-01  15:22:00             M          1939.0        76.0   

      driver_race                   violation_raw         violation  \
42160       Black         Other Traffic Violation  Moving violation   
74882       Black                        Speeding          Speeding   
87208       Black  Equipment/Inspection Violation         Equipment   
91154       Black                Call for Service             Other   

       search_conducted search_type stop_outcome is_arrested stop_duration  \
42160             False         NaN     Citation       False      0-15 Min   
74882             False         NaN     Citation       False      0-15 Min   
87208             False         NaN     Citation       False      0-15 Min   
91154             False         NaN    No Action       False     16-30 Min   

       drugs_related_stop  
42160               False  
74882               False  
87208               False  
91154               False

# set_index()
- So, we can use Pandas “set_index()” function to set any of your columns as the index.

In [265]:
df.set_index('driver_race')[4:9]

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
driver_race                                                                  
White       2005-03-14  10:00:00             F          1984.0        21.0   
Black       2005-03-23  09:45:00             M          1982.0        23.0   
White       2005-04-01  17:30:00             M          1969.0        36.0   
White       2005-06-06  13:20:00             F          1986.0        19.0   
Black       2005-07-13  10:15:00             M          1970.0        35.0   

                              violation_raw  violation  search_conducted  \
driver_race                                                                
White                              Speeding   Speeding             False   
Black        Equipment/Inspection Violation  Equipment             False   
White                              Speeding   Speeding             False   
White                              Speeding   Speeding             False   
Black                              Speeding   Speeding             False   

            search_type stop_outcome is_arrested stop_duration  \
driver_race                                                      
White               NaN     Citation       False      0-15 Min   
Black               NaN     Citation       False      0-15 Min   
White               NaN     Citation       False      0-15 Min   
White               NaN     Citation       False      0-15 Min   
Black               NaN     Citation       False      0-15 Min   

             drugs_related_stop  
driver_race                      
White                     False  
Black                     False  
White                     False  
White                     False  
Black                     False

# duplicated()
- duplicated() function to find all the duplicate rows in our dataset. 
- we can remove duplicate values using the drop_duplicates() function, 

as having too many duplicate values will affect the accuracy of our model at the later stage.

In [266]:
df.duplicated().sum()

479

In [267]:
df.drop_duplicates(inplace=True)

In [268]:
df.duplicated().sum()

0

In [269]:
df.shape

(91262, 14)

In [270]:
91741-479                                             #479 rows are droped 

91262

# get_dummies()
- get_dummies() method converts the categorical features of the data into dummy variables or indicator variables
- We usually do this conversion because some machine learning models don’t work well with categorical values such as Random Forrest, but we shouldn’t be using this if we have too many categories in our feature as it will create that many new features in our data frame, which will have an effect on the performance of our model.

#####   get_dummies()  method in Pandas provides a convenient way to perform One-Hot Encoding on data. 

In [271]:
df[:3]

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0 2005-01-02  01:55:00             M          1985.0        20.0       White   
1 2005-01-18  08:15:00             M          1965.0        40.0       White   
2 2005-01-23  23:15:00             M          1972.0        33.0       White   

  violation_raw violation  search_conducted search_type stop_outcome  \
0      Speeding  Speeding             False         NaN     Citation   
1      Speeding  Speeding             False         NaN     Citation   
2      Speeding  Speeding             False         NaN     Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False

In [272]:
df1 = pd.get_dummies(df.driver_race, prefix = 'driver_race_').iloc[:,0:]

In [273]:
df1.shape

(91262, 5)

# select_dtypes()
- Separate NUMARICAL and CATEGORICAL features from our data frame
- We can Create new ones by using the “select_dtypes()” function and include  [np.number]  to select numerical columns
- And include “objects” for categorical columns.

In [276]:
df2= df.select_dtypes(include=[np.number])

df2[4:8]

driver_age_raw  driver_age
4          1984.0        21.0
5          1982.0        23.0
6          1969.0        36.0
7          1986.0        19.0

In [277]:
df3 = df.select_dtypes(include='object')

df3[:5]

stop_time driver_gender driver_race     violation_raw violation search_type  \
0  01:55:00             M       White          Speeding  Speeding         NaN   
1  08:15:00             M       White          Speeding  Speeding         NaN   
2  23:15:00             M       White          Speeding  Speeding         NaN   
3  17:15:00             M       White  Call for Service     Other         NaN   
4  10:00:00             F       White          Speeding  Speeding         NaN   

    stop_outcome is_arrested stop_duration  
0       Citation       False      0-15 Min  
1       Citation       False      0-15 Min  
2       Citation       False      0-15 Min  
3  Arrest Driver        True     16-30 Min  
4       Citation       False      0-15 Min

# concat()
- We can perform concatenation of pandas object into a DataFrame output along a particular axis with optional set logic such as union and intersection using concat() method.

- By default, axis=0, i.e. row-wise concatenation, so if we set axis=1, column-wise concatenation will be performed.

In [279]:
df_combo = pd.concat([df2,df3],axis=1)

In [280]:
df_combo[:3]

driver_age_raw  driver_age stop_time driver_gender driver_race  \
0          1985.0        20.0  01:55:00             M       White   
1          1965.0        40.0  08:15:00             M       White   
2          1972.0        33.0  23:15:00             M       White   

  violation_raw violation search_type stop_outcome is_arrested stop_duration  
0      Speeding  Speeding         NaN     Citation       False      0-15 Min  
1      Speeding  Speeding         NaN     Citation       False      0-15 Min  
2      Speeding  Speeding         NaN     Citation       False      0-15 Min

# apply()
- while applying self defined function to your dataframe
- apply() :allows us to apply a custom function to every element of a particular Series.

In [282]:
def CurrentAgeCalculator(dob):
    return (2023-dob)


df['driver_current_age'] = df['driver_age_raw'].apply(CurrentAgeCalculator)

In [286]:
df['driver_current_age'].head()

0    38.0
1    58.0
2    51.0
3    37.0
4    39.0
Name: driver_current_age, dtype: float64

# to_csv()
- To save our organized DataFrame in a CSV file using the pandas “to_csv()” function. 
- As we don’t want to store the preceding indices of each row, hence, we will set index=False.

In [287]:
df_combo.to_csv("output.csv", index = False)

In [288]:
df4 = pd.read_csv("output.csv")

In [289]:
df4.columns

Index(['driver_age_raw', 'driver_age', 'stop_time', 'driver_gender',
       'driver_race', 'violation_raw', 'violation', 'search_type',
       'stop_outcome', 'is_arrested', 'stop_duration'],
      dtype='object')

In [290]:
df4.head(4)

driver_age_raw  driver_age stop_time driver_gender driver_race  \
0          1985.0        20.0  01:55:00             M       White   
1          1965.0        40.0  08:15:00             M       White   
2          1972.0        33.0  23:15:00             M       White   
3          1986.0        19.0  17:15:00             M       White   

      violation_raw violation search_type   stop_outcome is_arrested  \
0          Speeding  Speeding         NaN       Citation       False   
1          Speeding  Speeding         NaN       Citation       False   
2          Speeding  Speeding         NaN       Citation       False   
3  Call for Service     Other         NaN  Arrest Driver        True   

  stop_duration  
0      0-15 Min  
1      0-15 Min  
2      0-15 Min  
3     16-30 Min

# Thank you!
Faizan Shakeel | Aliagrh Muslim University | 8 June 2023